### DISCUSSION TITLE
Search directly in orui

### DISCUSSION THREAD

**Discussion by edumerco** [org-roam/org-roam-ui#85] [Feature request]
Dear @ThomasFKJorna, @kirillrogovoy and all the contributors. 

First things first, thank you very much for your work. I was scared to migrate to v2 mostly because I would loose the visualization, and your work with org-roam-ui made that worry disapear. :)

Now, I'm sorry if it's obvious for everyone except me, but I searched and didn't found how/if it's possible to search the nodes from within orui. Is it possible? :sweat_smile: 

If not, just a search in the titles would be a great start. And again, sorry if this is super clear for everyone.

Take care and have a great time... :smile:

**Comment by tefkah** [comment]
Thanks for your kind words! No, at the moment it's not possible, because for this first version we wanted to keep the things that Emacs does best in Emacs, which included org-roam-node-find. The best alternative we have now is turning on org-roam-ui-follow-mode, but we will definitely add this in the feature!

**Reply by edumerco** [reply]
Ah, completely understood and agreed. 
Actually, if you ever add this feature I can help testing it. :smirk:  
Thanks again and take care...

**Comment by kirillrogovoy** [comment]
Hey Eduardo!

Thanks for sharing the idea!

Could you, please also describe your workflow where this feature makes a serious difference?

E.g. I'm a very heavy user of org-roam-node-find and I can't see an example among my own use cases where I'd want to have the search inside org-roam-ui. It'd be great to understand the whole picture. :)

**Reply by tefkah** [reply]
Yes good question! I could see myself not per se wanting to find the node in emacs first (I have so many layers of customization that opening large buffers with e.g. a lot of LaTeX in it can take a while) while I would like to see the connections, maybe. 
More in general, I think that being able to explore your notes solely from the UI is valuable in it's own right.

**Reply by edumerco** [reply]
Hello @kirillrogovoy and @ThomasFKJorna.

> Could you, please also describe your workflow where this feature makes a serious difference?

It will be a pleasure, and please forgive my non native English. :)

For me, understanding connections is fundamental. As many other people I do it more in a visual way (orui) than by text (org-roam backlinks buffer or by list completion). This is so much so that in 2007 with a friend we did something similar in MoinMoin (see https://moinmo.in/VisualSiteMap). :)

In our present case (assuming a couple hundred or more nodes) we start with a node and relationships map that is not understandable directly, since there are just too many node and relationships, and names only appear when we zoom in, so currently it is possible to do a visual exploration by hovering en orui but we can only do that node by node, one at a time.

As an example, let's imagine that I'm looking for nodes about sustainab* nodes. I can do a search in emacs (I have helm here, but other completion frameworks are the same for our example) and select a node, and then (only after a selection) I'll see it in orui. In my case "sust" gives me 6 nodes and 3 or 4 more aliases for those nodes because of their translations. In order to see that region, I am forced to select one (and only one) and go to orui. And then I'll only see that one and it's 1st steps, not the whole region. 

Wouldn't it be great if I can do that search in orui and see the nodes and their relationships, just as if I here hovering over all of them at the same time? In this (imaginary) example when I input "sust" I'd see those 6 nodes listed in the search completion iu and in the graph just the same as if I hover over them at the same time so I can visualize the topology of the "sust" region (something impossible today since hovering is a one-by-one action). Also, this comes with an almost free reduction in clutter, since those aliases (extra data) are for the nodes already highlighted, so I see only 6 of them. 

And this new ability becomes even better with 2 or more terms.

Let's look to "inference" in other part of the same concept map. Now, only 2 nodes appear in the list: "ladder of inference" and "inference types". Currently I cannot explore if/how both regions (sust and inference) have a relation. But if orui allows me to search I can enter "inference OR sust" and I would see highlighted in the graph where these regions are and/if they are near or far, or do have connections or some overlap among them.

In every case, selecting a node from the list (I do imagine it very much like helm, typing more makes the list shorter and more specific and selecting one is done by Enter or click) goes there like it is today when we click in a node (this is, the default orui action).

I hope this is more or less explained, but in any case, please do ask for clarifications and I'll be happy to ella/colla/borate. :)

Thank you again for your great work. :)
Best...

**Reply by kirillrogovoy** [reply]
Thank you!

One thing I'm thinking right off the bat is that there seems to be some way to add actions to be executed in the context of a Helm list. Looking at it right now, it seems like the only action that exists is just navigating to that node, but maybe we could add one or a few more that would tell org-roam-ui to narrow the scope of the graph.

I like the idea overall, but I'm struggling to give up the perfectly working org-roam-node-find just to re-implement it in the web app. But if it could be extended, it should have probably been more or less trivial to this set of features, especially since we already have an internal state for narrowing down the scope (aka the "local mode").

**Reply by edumerco** [reply]
> One thing I'm thinking right off the bat is that there seems to be some way to add actions to be executed in the context of a Helm list. Looking at it right now, it seems like the only action that exists is just navigating to that node, but maybe we could add one or a few more that would tell org-roam-ui to narrow the scope of the graph.

That sounds perfect. Using emacs+orui side by side we can use helm for this. :)

The only issue I see there is that we are talking some not-so-small screen space (the emacs surface) only to search. But it's a great start. 

> I like the idea overall, but I'm struggling to give up the perfectly working org-roam-node-find just to re-implement it in the web app. But if it could be extended, it should have probably been more or less trivial to this set of features, 

Can we emulate the helm behavior but in orui "talking" with helm itself? Sorry if it is ridiculous, I don't know the internals of the implementation, or it's possibilities...

> especially since we already have an internal state for narrowing down the scope (aka the "local mode").

That's great! :D

**Reply by tefkah** [reply]
I actually really like the idea of "dynamically filtering" the nodes by first highlighting them, and then being able to "select" them by going into local mode. 

I think it shouldn't be too hard to send the info fetched by org-roam-node-find to ORUI, but I don't think we should make that Helm/Ivy/Selectrum/whatever dependent as that will be a nightmare. I don't really know much about those completion frameworks, but they fetch that info somehow, and I assume they do it similarly, so we could probably hook into it.

However, I would probably still like to have the option of doing this in the webapp, kind of like how Obsidian does it, *but* in combination with generating a list of entries you can click on (and ideally preview, #64) in a sidebar. These should work complementarily though, and that needs some thinkin'.

**Comment by Dspil** [comment]
Hello everyone,

Thank you for this great tool. That feature seems useful to me because I am using org roam ui in my mobile phone where I don't really use emacs, just to find something on the go. There a search function would be extremely useful
